File: code/geometry/intersect_cc.py
def intersectCC(c1, c2, r1, r2):
	if c1 == c2:
		assert(r1 != r2)
		return None
	vec = vecsub(c2, c1)
	d2, sm, dif = len2(vec), r1 + r2, r1 - r2
	if sm ** 2 < d2 or dif ** 2 > d2:
		return None
	p = (d2 + r1 ** 2 - r2 ** 2) / (d2 * 2)
	h2 = r1 ** 2 - p * p * d2
	mid = (c1[0] + vec[0] * p, c1[1] + vec[1] * p)
	plen = sqrt(max(0, h2) / d2)
	per = (-vec[1] * plen, vec[0] * plen)
	return (vecadd(mid, per), vecsub(mid, per))

File: code/geometry_cpp/intersect_cc.cpp
template <typename T> optional<array<dpoint, 2>>
intersectCC(point<T> c1, point<T> c2, T r1, T r2) {
	if (c1 == c2) {
		assert(r1 != r2);
		return { };
	}
	auto vec = c2 - c1;
	T d2 = vec.len2(), sm = r1 + r2, dif = r1 - r2;
	if (sm * sm < d2 || dif * dif > d2)
		return { };
	double p = double(d2 + r1*r1 - r2*r2) / (d2 * 2);
	dpoint mid(c1.x + vec.x * p, c1.y + vec.y * p);
	dpoint per = dpoint(-vec.y, vec.x) * sqrt(max(0.0, r1 * r1 - p * p * d2) / d2);
	return { { mid+per, mid-per } };
}
